File: src/restork/daily/apple_music.py
"""Official Apple Music API public catalog-playlist adapter."""

from __future__ import annotations

import asyncio
import json
import re
from collections.abc import Callable, Mapping
from datetime import UTC, date, datetime
from hashlib import sha256
from urllib.parse import parse_qsl, quote, urljoin, urlsplit, urlunsplit

from restork.contracts.outbound import OutboundEnvelope
from restork.contracts.types import DataClass, PolicyDecision
from restork.daily.music import PlaylistDocument, PlaylistItem, PlaylistSource
from restork.network.gateway import (
    OutboundDeniedError,
    OutboundGateway,
    OutboundRequest,
    OutboundResponse,
)

_API_ORIGIN = "https://api.music.apple.com"
_ARTWORK_HOSTS = frozenset(
    {
        "is1-ssl.mzstatic.com",
        "is2-ssl.mzstatic.com",
        "is3-ssl.mzstatic.com",
        "is4-ssl.mzstatic.com",
        "is5-ssl.mzstatic.com",
    }
)
_APPLE_ID = re.compile(r"^[A-Za-z0-9._-]{3,128}$")
_MAXIMUM_TRACKS = 2_000
_MAXIMUM_PAGES = 20


class AppleMusicError(RuntimeError):
    """A bounded official-provider failure that never includes credential material."""


class AppleMusicClient:
    def __init__(
        self,
        gateway: OutboundGateway,
        developer_token: Callable[[], str],
        music_user_token: Callable[[], str] | None = None,
    ) -> None:
        self._gateway = gateway
        self._developer_token = developer_token
        self._music_user_token = music_user_token
        self._cover_cache: dict[str, tuple[bytes, str]] = {}

    async def synchronize(
        self,
        share_url: str,
        *,
        on_date: date,
        preferred_genres: tuple[str, ...] = (),
    ) -> PlaylistDocument:
        del on_date, preferred_genres
        storefront, playlist_id, public_url = parse_apple_music_playlist_url(share_url)
        return await self._synchronize_identity(storefront, playlist_id, public_url)

    async def synchronize_id(
        self,
        source_id: str,
        *,
        on_date: date | None = None,
        preferred_genres: tuple[str, ...] = (),
    ) -> PlaylistDocument:
        del on_date, preferred_genres
        storefront, separator, playlist_id = source_id.partition(":")
        if (
            not separator
            or len(storefront) != 2
            or not storefront.isascii()
            or not storefront.islower()
            or not storefront.isalpha()
            or _APPLE_ID.fullmatch(playlist_id) is None
        ):
            raise ValueError("Apple Music playlist identity is invalid.")
        return await self._synchronize_identity(storefront, playlist_id, "")

    async def _synchronize_identity(
        self, storefront: str, playlist_id: str, submitted_public_url: str
    ) -> PlaylistDocument:
        developer_token = await asyncio.to_thread(self._developer_token)
        _validate_secret(developer_token)
        music_user_token = ""
        if self._music_user_token is not None:
            try:
                music_user_token = await asyncio.to_thread(self._music_user_token)
                _validate_secret(music_user_token)
            except LookupError:
                music_user_token = ""
        headers = {
            "Accept": "application/json",
            "Authorization": f"Bearer {developer_token}",
        }
        if music_user_token:
            headers["Music-User-Token"] = music_user_token
        endpoint = (
            f"{_API_ORIGIN}/v1/catalog/{storefront}/playlists/"
            f"{quote(playlist_id, safe='._-')}?include=tracks"
        )
        root = _json_response(
            await self._dispatch(
                endpoint,
                purpose="daily_music_apple_catalog_playlist_sync",
                source_refs=(f"apple-playlist:{sha256(playlist_id.encode()).hexdigest()[:24]}",),
                headers=headers,
            )
        )
        data = root.get("data")
        if not isinstance(data, list) or len(data) != 1:
            raise AppleMusicError("Apple Music playlist response is incomplete.")
        playlist = _object(data[0], "playlist")
        if (
            _text(playlist.get("id"), "playlist ID", maximum=128) != playlist_id
            or playlist.get("type") != "playlists"
        ):
            raise AppleMusicError("Apple Music response does not match the requested playlist.")
        attributes = _object(playlist.get("attributes"), "playlist attributes")
        title = _text(attributes.get("name"), "playlist title", maximum=300)
        provider_public_url = _public_url(attributes.get("url"), required=False)
        public_url = provider_public_url or submitted_public_url
        if not public_url:
            raise AppleMusicError("Apple Music playlist URL is unavailable.")
        relationships = _object(playlist.get("relationships"), "playlist relationships")
        page = _object(relationships.get("tracks"), "playlist tracks")
        raw_tracks = _track_list(page.get("data"))
        next_url = page.get("next")
        pages = 1
        while next_url is not None:
            if pages >= _MAXIMUM_PAGES or len(raw_tracks) >= _MAXIMUM_TRACKS:
                raise AppleMusicError("Apple Music playlist pagination exceeds its bounds.")
            selected_url = _next_url(next_url, storefront, playlist_id)
            page = _json_response(
                await self._dispatch(
                    selected_url,
                    purpose="daily_music_apple_catalog_playlist_page",
                    source_refs=(
                        f"apple-playlist:{sha256(playlist_id.encode()).hexdigest()[:24]}",
                    ),
                    headers=headers,
                )
            )
            raw_tracks.extend(_track_list(page.get("data")))
            next_url = page.get("next")
            pages += 1
        if not raw_tracks or len(raw_tracks) > _MAXIMUM_TRACKS:
            raise AppleMusicError("Apple Music playlist contains no bounded readable tracks.")
        items = tuple(_track(track, storefront) for track in raw_tracks)
        identities = [item.item_id for item in items]
        if len(identities) != len(set(identities)):
            raise AppleMusicError("Apple Music playlist contains duplicate track identifiers.")
        return PlaylistDocument(
            items=items,
            source=PlaylistSource(
                provider="apple-music",
                source_id=f"{storefront}:{playlist_id}",
                label=title,
                public_url=public_url,
                synced_at=datetime.now(UTC),
                refresh_supported=True,
                experimental=False,
                official_api=True,
                read_only=True,
                requires_user_consent=True,
                supports_charts=False,
            ),
        )

    async def fetch_cover(self, url: str) -> tuple[bytes, str]:
        selected = _artwork_url(url)
        cached = self._cover_cache.get(selected)
        if cached is not None:
            return cached
        response = await self._dispatch(
            selected,
            purpose="daily_music_apple_cover",
            source_refs=(f"apple-cover:{sha256(selected.encode()).hexdigest()[:24]}",),
            headers={"Accept": "image/jpeg,image/png,image/webp"},
        )
        if response.status_code != 200:
            raise AppleMusicError("Apple Music cover returned an error.")
        media_type = _media_type(response.headers)
        if media_type not in {"image/jpeg", "image/png", "image/webp"} or not _valid_image(
            response.payload, media_type
        ):
            raise AppleMusicError("Apple Music cover returned unsupported image data.")
        if len(self._cover_cache) >= 8:
            self._cover_cache.pop(next(iter(self._cover_cache)))
        self._cover_cache[selected] = response.payload, media_type
        return response.payload, media_type

    async def _dispatch(
        self,
        destination: str,
        *,
        purpose: str,
        source_refs: tuple[str, ...],
        headers: Mapping[str, str],
    ) -> OutboundResponse:
        envelope = OutboundEnvelope(
            destination=destination,
            resolved_address_class="public",
            method="GET",
            purpose=purpose,
            source_refs=list(source_refs),
            payload_hash=sha256(b"").hexdigest(),
            classification=DataClass.PERSONAL,
            redaction_summary=(
                "official Apple Music request; credentials remain in ephemeral headers and "
                "are excluded from the audit envelope"
            ),
            policy_version="v1",
            policy_decision=PolicyDecision.ALLOWED,
        )
        try:
            return await self._gateway.dispatch(
                OutboundRequest(
                    envelope=envelope,
                    payload=b"",
                    headers={
                        "User-Agent": "Restork/0.1 (+https://github.com/Totoro-qaq/restork)",
                        **headers,
                    },
                )
            )
        except OutboundDeniedError as error:
            raise AppleMusicError("Apple Music request was denied by outbound policy.") from error


def parse_apple_music_playlist_url(value: str) -> tuple[str, str, str]:
    normalized = value.strip()
    if not normalized or len(normalized) > 2_048:
        raise ValueError("Apple Music playlist share link is empty or too long.")
    try:
        parsed = urlsplit(normalized)
        port = parsed.port
    except ValueError as error:
        raise ValueError("Apple Music playlist share link is invalid.") from error
    segments = [segment for segment in parsed.path.split("/") if segment]
    if (
        parsed.scheme != "https"
        or parsed.hostname != "music.apple.com"
        or parsed.username is not None
        or parsed.password is not None
        or parsed.fragment
        or port not in {None, 443}
        or len(segments) != 4
        or segments[1] != "playlist"
        or len(segments[0]) != 2
        or not segments[0].isascii()
        or not segments[0].isalpha()
        or not segments[2]
        or len(segments[2]) > 200
        or _APPLE_ID.fullmatch(segments[3]) is None
    ):
        raise ValueError("Use a public Apple Music catalog-playlist link.")
    storefront = segments[0].lower()
    playlist_id = segments[3]
    public_url = (
        f"https://music.apple.com/{storefront}/playlist/"
        f"{quote(segments[2], safe='._-')}/{quote(playlist_id, safe='._-')}"
    )
    return storefront, playlist_id, public_url


def _track(value: object, storefront: str) -> PlaylistItem:
    track = _object(value, "track")
    track_id = _text(track.get("id"), "track ID", maximum=128)
    if track.get("type") != "songs" or _APPLE_ID.fullmatch(track_id) is None:
        raise AppleMusicError("Apple Music track identity is invalid.")
    attributes = _object(track.get("attributes"), "track attributes")
    genres = attributes.get("genreNames", [])
    if not isinstance(genres, list) or len(genres) > 20:
        raise AppleMusicError("Apple Music genres are invalid.")
    selected_genres = tuple(
        _text(genre, "genre", maximum=80) for genre in genres[:5]
    )
    published_on = _optional_date(attributes.get("releaseDate"))
    facts = [
        f"released {published_on.isoformat()}" if published_on else "",
        f"genre: {' / '.join(selected_genres)}" if selected_genres else "",
    ]
    analysis = "; ".join(fact for fact in facts if fact)
    artwork = attributes.get("artwork")
    cover_url = (
        _artwork_url(_object(artwork, "artwork").get("url")) if artwork is not None else ""
    )
    source_url = _public_url(attributes.get("url"), required=False) or (
        f"https://music.apple.com/{storefront}/song/{quote(track_id, safe='._-')}"
    )
    return PlaylistItem(
        item_id=f"apple-music:{track_id}",
        title=_text(attributes.get("name"), "track name", maximum=300),
        artist=_text(attributes.get("artistName"), "artist name", maximum=300),
        album=_text(
            attributes.get("albumName", ""), "album name", maximum=300, required=False
        ),
        tags=("apple-music", *selected_genres),
        note=(
            f"Apple Music catalog metadata records {analysis}."
            if analysis
            else "No reviewed structured song metadata is available."
        ),
        cover_url=cover_url,
        source_provider="apple-music",
        source_item_id=track_id,
        source_url=source_url,
        genre=" / ".join(selected_genres),
        published_on=published_on,
    )


def _next_url(value: object, storefront: str, playlist_id: str) -> str:
    if not isinstance(value, str) or not value:
        raise AppleMusicError("Apple Music pagination URL is invalid.")
    selected = urljoin(f"{_API_ORIGIN}/", value)
    parsed = urlsplit(selected)
    expected_path = (
        f"/v1/catalog/{storefront}/playlists/{quote(playlist_id, safe='._-')}/tracks"
    )
    if (
        parsed.scheme != "https"
        or parsed.hostname != "api.music.apple.com"
        or parsed.username is not None
        or parsed.password is not None
        or parsed.fragment
        or parsed.path != expected_path
        or any(key not in {"offset", "limit"} for key, _ in parse_qsl(parsed.query))
    ):
        raise AppleMusicError("Apple Music pagination URL is invalid.")
    return selected


def _public_url(value: object, *, required: bool = True) -> str:
    if value in {None, ""} and not required:
        return ""
    if not isinstance(value, str) or len(value) > 2_048:
        raise AppleMusicError("Apple Music public URL is invalid.")
    parsed = urlsplit(value)
    if (
        parsed.scheme != "https"
        or parsed.hostname != "music.apple.com"
        or parsed.username is not None
        or parsed.password is not None
        or parsed.fragment
    ):
        raise AppleMusicError("Apple Music public URL is invalid.")
    return urlunsplit((parsed.scheme, parsed.netloc, parsed.path, "", ""))


def _artwork_url(value: object) -> str:
    if not isinstance(value, str) or not value or len(value) > 4_096:
        raise AppleMusicError("Apple Music artwork URL is invalid.")
    rendered = value.replace("{w}", "300").replace("{h}", "300").replace("{f}", "jpg")
    parsed = urlsplit(rendered)
    if (
        parsed.scheme != "https"
        or parsed.hostname not in _ARTWORK_HOSTS
        or parsed.username is not None
        or parsed.password is not None
        or parsed.fragment
        or "{" in rendered
        or not parsed.path.lower().endswith((".jpg", ".jpeg", ".png", ".webp"))
    ):
        raise AppleMusicError("Apple Music artwork URL is invalid.")
    return rendered


def _validate_secret(value: str) -> None:
    if (
        not value
        or len(value) > 16_384
        or any(character.isspace() or ord(character) < 32 for character in value)
    ):
        raise AppleMusicError("Apple Music native credential is invalid.")


def _json_response(response: OutboundResponse) -> dict[str, object]:
    if response.status_code != 200:
        raise AppleMusicError("Apple Music API returned an error.")
    try:
        value = json.loads(response.payload)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise AppleMusicError("Apple Music API returned invalid JSON.") from error
    return _object(value, "response")


def _object(value: object, label: str) -> dict[str, object]:
    if not isinstance(value, dict) or not all(isinstance(key, str) for key in value):
        raise AppleMusicError(f"Apple Music {label} is invalid.")
    return value


def _track_list(value: object) -> list[object]:
    if not isinstance(value, list):
        raise AppleMusicError("Apple Music track page is invalid.")
    return list(value)


def _text(value: object, label: str, *, maximum: int, required: bool = True) -> str:
    if not isinstance(value, str):
        raise AppleMusicError(f"Apple Music {label} is invalid.")
    selected = " ".join(value.split())
    if (required and not selected) or len(selected) > maximum:
        raise AppleMusicError(f"Apple Music {label} is invalid.")
    return selected


def _optional_date(value: object) -> date | None:
    if value in {None, ""}:
        return None
    if not isinstance(value, str) or len(value) != 10:
        return None
    try:
        return date.fromisoformat(value)
    except ValueError:
        return None


def _media_type(headers: Mapping[str, str]) -> str:
    value = next(
        (item for key, item in headers.items() if key.casefold() == "content-type"),
        "",
    )
    return value.split(";", 1)[0].strip().casefold()


def _valid_image(payload: bytes, media_type: str) -> bool:
    if media_type == "image/jpeg":
        return payload.startswith(b"\xff\xd8\xff")
    if media_type == "image/png":
        return payload.startswith(b"\x89PNG\r\n\x1a\n")
    return payload.startswith(b"RIFF") and payload[8:12] == b"WEBP"


__all__ = ["AppleMusicClient", "AppleMusicError", "parse_apple_music_playlist_url"]

File: rust/crates/restork-daily/src/apple_music.rs
//! Official Apple Music API catalog-playlist connector.
//!
//! Credentials are supplied by Core after native-secret resolution and are never serialized.
//! The adapter does not scrape Apple Music pages or attempt private-library access.

use std::collections::BTreeSet;

use chrono::{NaiveDate, Utc};
use reqwest::header::{self, HeaderValue};
use serde::Deserialize;
use url::Url;

use super::{
    DailyClient, DailyError, MusicSourceDocument, MusicSourceSummary, PlaylistItem,
    bounded_json_with_limit, music_snapshot, validate_playlist_item,
};
use crate::qqmusic::bounded_image;

const API_ORIGIN: &str = "https://api.music.apple.com";
const MAX_TRACKS: usize = 2_000;
const MAX_PAGES: usize = 20;
const MAX_PROVIDER_RESPONSE_BYTES: usize = 4_000_000;

pub type AppleMusicDocument = MusicSourceDocument;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AppleMusicPlaylistIdentity {
    pub storefront: String,
    pub playlist_id: String,
    pub public_url: String,
}

/// Normalize a public Apple Music catalog-playlist link without retaining tracking parameters.
pub fn parse_apple_music_playlist(
    share_url: &str,
) -> Result<AppleMusicPlaylistIdentity, DailyError> {
    if share_url.is_empty() || share_url.len() > 2_048 {
        return Err(DailyError::InvalidInput);
    }
    let parsed = Url::parse(share_url).map_err(|_| DailyError::InvalidInput)?;
    if parsed.scheme() != "https"
        || parsed.host_str() != Some("music.apple.com")
        || !parsed.username().is_empty()
        || parsed.password().is_some()
        || parsed.port_or_known_default() != Some(443)
        || parsed.fragment().is_some()
    {
        return Err(DailyError::InvalidInput);
    }
    let segments = parsed
        .path_segments()
        .map(|segments| segments.filter(|part| !part.is_empty()).collect::<Vec<_>>())
        .unwrap_or_default();
    let [storefront, "playlist", slug, playlist_id] = segments.as_slice() else {
        return Err(DailyError::InvalidInput);
    };
    if storefront.len() != 2
        || !storefront.bytes().all(|byte| byte.is_ascii_alphabetic())
        || slug.is_empty()
        || slug.chars().count() > 200
        || !valid_apple_id(playlist_id)
    {
        return Err(DailyError::InvalidInput);
    }
    let storefront = storefront.to_ascii_lowercase();
    Ok(AppleMusicPlaylistIdentity {
        storefront: storefront.clone(),
        playlist_id: (*playlist_id).to_owned(),
        public_url: format!(
            "https://music.apple.com/{storefront}/playlist/{}/{}",
            encode_path_segment(slug),
            encode_path_segment(playlist_id)
        ),
    })
}

impl DailyClient {
    pub async fn sync_apple_music(
        &self,
        share_url: &str,
        local_date: &str,
        developer_token: &str,
        music_user_token: Option<&str>,
    ) -> Result<AppleMusicDocument, DailyError> {
        let identity = parse_apple_music_playlist(share_url)?;
        self.sync_apple_music_identity(
            &identity.storefront,
            &identity.playlist_id,
            &identity.public_url,
            local_date,
            developer_token,
            music_user_token,
        )
        .await
    }

    pub async fn sync_apple_music_id(
        &self,
        source_identity: &str,
        local_date: &str,
        developer_token: &str,
        music_user_token: Option<&str>,
    ) -> Result<AppleMusicDocument, DailyError> {
        let (storefront, playlist_id) = source_identity
            .split_once(':')
            .ok_or(DailyError::InvalidInput)?;
        if storefront.len() != 2
            || !storefront.bytes().all(|byte| byte.is_ascii_lowercase())
            || !valid_apple_id(playlist_id)
        {
            return Err(DailyError::InvalidInput);
        }
        self.sync_apple_music_identity(
            storefront,
            playlist_id,
            "",
            local_date,
            developer_token,
            music_user_token,
        )
        .await
    }

    async fn sync_apple_music_identity(
        &self,
        storefront: &str,
        playlist_id: &str,
        submitted_public_url: &str,
        local_date: &str,
        developer_token: &str,
        music_user_token: Option<&str>,
    ) -> Result<AppleMusicDocument, DailyError> {
        if NaiveDate::parse_from_str(local_date, "%Y-%m-%d").is_err() {
            return Err(DailyError::InvalidInput);
        }
        let authorization = bearer_header(developer_token)?;
        let user_token = music_user_token.map(secret_header).transpose()?;
        let endpoint = format!(
            "{API_ORIGIN}/v1/catalog/{storefront}/playlists/{}",
            encode_path_segment(playlist_id)
        );
        let mut request = self
            .client
            .get(endpoint)
            .query(&[("include", "tracks")])
            .header(header::ACCEPT, "application/json")
            .header(header::AUTHORIZATION, authorization.clone());
        if let Some(token) = user_token.as_ref() {
            request = request.header("Music-User-Token", token.clone());
        }
        let response = request.send().await.map_err(|_| DailyError::Unavailable)?;
        let root: CatalogResponse =
            bounded_json_with_limit(response, MAX_PROVIDER_RESPONSE_BYTES).await?;
        if root.data.len() != 1 {
            return Err(DailyError::InvalidResponse);
        }
        let playlist = root
            .data
            .into_iter()
            .next()
            .ok_or(DailyError::InvalidResponse)?;
        if playlist.id != playlist_id || playlist.resource_type != "playlists" {
            return Err(DailyError::InvalidResponse);
        }
        let label = remote_text(&playlist.attributes.name, 300)?;
        let public_url = normalize_public_url(&playlist.attributes.url)
            .or_else(|| (!submitted_public_url.is_empty()).then(|| submitted_public_url.to_owned()))
            .ok_or(DailyError::InvalidResponse)?;
        let tracks = playlist
            .relationships
            .and_then(|relationships| relationships.tracks)
            .ok_or(DailyError::InvalidResponse)?;
        let mut all_tracks = tracks.data;
        let mut next = tracks.next;
        let mut pages = 1;
        while let Some(next_path) = next {
            if pages >= MAX_PAGES || all_tracks.len() >= MAX_TRACKS {
                return Err(DailyError::InvalidResponse);
            }
            let next_url = validate_next_url(&next_path, storefront, playlist_id)?;
            let mut request = self
                .client
                .get(next_url)
                .header(header::ACCEPT, "application/json")
                .header(header::AUTHORIZATION, authorization.clone());
            if let Some(token) = user_token.as_ref() {
                request = request.header("Music-User-Token", token.clone());
            }
            let response = request.send().await.map_err(|_| DailyError::Unavailable)?;
            let page: TrackPage =
                bounded_json_with_limit(response, MAX_PROVIDER_RESPONSE_BYTES).await?;
            all_tracks.extend(page.data);
            next = page.next;
            pages += 1;
        }
        let items = normalize_tracks(all_tracks, storefront)?;
        if music_snapshot(&items, local_date).recommendation.is_none() {
            return Err(DailyError::InvalidResponse);
        }
        Ok(AppleMusicDocument {
            provider: "apple-music".to_owned(),
            source_identity: format!("{storefront}:{playlist_id}"),
            source: MusicSourceSummary {
                provider: "apple-music".to_owned(),
                label,
                item_count: items.len(),
                synced_at: Some(Utc::now().to_rfc3339()),
                public_url,
                refresh_supported: true,
                experimental: false,
                official_api: true,
                read_only: true,
                requires_user_consent: true,
                supports_charts: false,
            },
            items,
            discoveries: Vec::new(),
        })
    }

    pub async fn apple_music_cover(
        &self,
        cover_url: &str,
    ) -> Result<(Vec<u8>, String), DailyError> {
        let cover_url = normalize_artwork_url(cover_url)?;
        let response = self
            .client
            .get(cover_url)
            .header(header::ACCEPT, "image/jpeg,image/png,image/webp")
            .header(header::REFERER, "https://music.apple.com/")
            .send()
            .await
            .map_err(|_| DailyError::Unavailable)?;
        bounded_image(response).await
    }
}

fn normalize_tracks(
    tracks: Vec<TrackResource>,
    storefront: &str,
) -> Result<Vec<PlaylistItem>, DailyError> {
    if tracks.is_empty() || tracks.len() > MAX_TRACKS {
        return Err(DailyError::InvalidResponse);
    }
    let mut identities = BTreeSet::new();
    let mut items = Vec::with_capacity(tracks.len());
    for track in tracks {
        if track.resource_type != "songs" || !valid_apple_id(&track.id) {
            return Err(DailyError::InvalidResponse);
        }
        let item_id = format!("apple-music-{}", track.id);
        if !identities.insert(item_id.clone()) {
            return Err(DailyError::InvalidResponse);
        }
        let attributes = track.attributes.ok_or(DailyError::InvalidResponse)?;
        let genres = attributes
            .genre_names
            .iter()
            .take(5)
            .map(|genre| remote_text(genre, 80))
            .collect::<Result<Vec<_>, _>>()?;
        let published_on = attributes
            .release_date
            .as_deref()
            .and_then(|value| NaiveDate::parse_from_str(value, "%Y-%m-%d").ok())
            .map(|date| date.to_string());
        let facts = [
            published_on.as_ref().map(|date| format!("released {date}")),
            (!genres.is_empty()).then(|| format!("genre: {}", genres.join(" / "))),
        ]
        .into_iter()
        .flatten()
        .collect::<Vec<_>>();
        let source_url = attributes
            .url
            .as_deref()
            .and_then(normalize_public_url)
            .unwrap_or_else(|| format!("https://music.apple.com/{storefront}/song/{}", track.id));
        let item = PlaylistItem {
            item_id,
            title: remote_text(&attributes.name, 300)?,
            artist: remote_text(&attributes.artist_name, 300)?,
            album: optional_text(&attributes.album_name, 300)?,
            tags: std::iter::once("apple-music".to_owned())
                .chain(genres.iter().cloned())
                .collect(),
            analysis: if facts.is_empty() {
                "No reviewed structured song metadata is available.".to_owned()
            } else {
                format!("Apple Music catalog metadata records {}.", facts.join("; "))
            },
            cover_url: attributes.artwork.as_ref().map_or_else(
                || Ok(String::new()),
                |artwork| normalize_artwork_url(&artwork.url),
            )?,
            source_provider: "apple-music".to_owned(),
            source_item_id: track.id,
            source_url,
            language: String::new(),
            genre: genres.join(" / "),
            published_on,
            popularity_reason: String::new(),
        };
        validate_playlist_item(&item).map_err(|_| DailyError::InvalidResponse)?;
        items.push(item);
    }
    Ok(items)
}

fn bearer_header(token: &str) -> Result<HeaderValue, DailyError> {
    let token = secret_header(token)?;
    HeaderValue::from_str(&format!(
        "Bearer {}",
        token.to_str().map_err(|_| DailyError::InvalidInput)?
    ))
    .map_err(|_| DailyError::InvalidInput)
}

fn secret_header(value: &str) -> Result<HeaderValue, DailyError> {
    if value.is_empty()
        || value.len() > 16_384
        || value
            .chars()
            .any(|character| character.is_control() || character.is_whitespace())
    {
        return Err(DailyError::InvalidInput);
    }
    HeaderValue::from_str(value).map_err(|_| DailyError::InvalidInput)
}

fn validate_next_url(value: &str, storefront: &str, playlist_id: &str) -> Result<Url, DailyError> {
    let parsed = if value.starts_with('/') {
        Url::parse(&format!("{API_ORIGIN}{value}"))
    } else {
        Url::parse(value)
    }
    .map_err(|_| DailyError::InvalidResponse)?;
    let expected_prefix = format!(
        "/v1/catalog/{storefront}/playlists/{}/tracks",
        encode_path_segment(playlist_id)
    );
    if parsed.scheme() != "https"
        || parsed.host_str() != Some("api.music.apple.com")
        || !parsed.username().is_empty()
        || parsed.password().is_some()
        || parsed.port_or_known_default() != Some(443)
        || parsed.fragment().is_some()
        || parsed.path() != expected_prefix
        || parsed
            .query_pairs()
            .any(|(key, _)| key != "offset" && key != "limit")
    {
        return Err(DailyError::InvalidResponse);
    }
    Ok(parsed)
}

fn normalize_public_url(value: &str) -> Option<String> {
    let mut parsed = Url::parse(value).ok()?;
    if parsed.scheme() != "https"
        || parsed.host_str() != Some("music.apple.com")
        || !parsed.username().is_empty()
        || parsed.password().is_some()
        || parsed.port_or_known_default() != Some(443)
        || parsed.fragment().is_some()
    {
        return None;
    }
    parsed.set_query(None);
    Some(parsed.into())
}

fn normalize_artwork_url(value: &str) -> Result<String, DailyError> {
    if value.is_empty() || value.len() > 4_096 {
        return Err(DailyError::InvalidResponse);
    }
    let rendered = value
        .replace("{w}", "300")
        .replace("{h}", "300")
        .replace("{f}", "jpg");
    let parsed = Url::parse(&rendered).map_err(|_| DailyError::InvalidResponse)?;
    if parsed.scheme() != "https"
        || !matches!(
            parsed.host_str(),
            Some(
                "is1-ssl.mzstatic.com"
                    | "is2-ssl.mzstatic.com"
                    | "is3-ssl.mzstatic.com"
                    | "is4-ssl.mzstatic.com"
                    | "is5-ssl.mzstatic.com"
            )
        )
        || !parsed.username().is_empty()
        || parsed.password().is_some()
        || parsed.port_or_known_default() != Some(443)
        || parsed.fragment().is_some()
        || rendered.contains('{')
        || !matches!(
            parsed
                .path()
                .rsplit_once('.')
                .map(|(_, extension)| extension),
            Some("jpg" | "jpeg" | "png" | "webp")
        )
    {
        return Err(DailyError::InvalidResponse);
    }
    Ok(parsed.into())
}

fn valid_apple_id(value: &str) -> bool {
    (3..=128).contains(&value.len())
        && value
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'-' | b'_'))
}

fn encode_path_segment(value: &str) -> String {
    url::form_urlencoded::byte_serialize(value.as_bytes()).collect()
}

fn remote_text(value: &str, maximum: usize) -> Result<String, DailyError> {
    let selected = value.split_whitespace().collect::<Vec<_>>().join(" ");
    if selected.is_empty()
        || selected.chars().count() > maximum
        || selected.chars().any(char::is_control)
    {
        Err(DailyError::InvalidResponse)
    } else {
        Ok(selected)
    }
}

fn optional_text(value: &str, maximum: usize) -> Result<String, DailyError> {
    if value.trim().is_empty() {
        Ok(String::new())
    } else {
        remote_text(value, maximum)
    }
}

#[derive(Deserialize)]
struct CatalogResponse {
    data: Vec<PlaylistResource>,
}

#[derive(Deserialize)]
struct PlaylistResource {
    id: String,
    #[serde(rename = "type")]
    resource_type: String,
    attributes: PlaylistAttributes,
    relationships: Option<PlaylistRelationships>,
}

#[derive(Deserialize)]
struct PlaylistAttributes {
    name: String,
    url: String,
}

#[derive(Deserialize)]
struct PlaylistRelationships {
    tracks: Option<TrackPage>,
}

#[derive(Deserialize)]
struct TrackPage {
    #[serde(default)]
    next: Option<String>,
    data: Vec<TrackResource>,
}

#[derive(Deserialize)]
struct TrackResource {
    id: String,
    #[serde(rename = "type")]
    resource_type: String,
    attributes: Option<TrackAttributes>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct TrackAttributes {
    name: String,
    artist_name: String,
    #[serde(default)]
    album_name: String,
    #[serde(default)]
    genre_names: Vec<String>,
    release_date: Option<String>,
    artwork: Option<Artwork>,
    url: Option<String>,
}

#[derive(Deserialize)]
struct Artwork {
    url: String,
}

#[cfg(test)]
mod tests {
    use super::{
        CatalogResponse, normalize_artwork_url, normalize_tracks, parse_apple_music_playlist,
        validate_next_url,
    };

    #[test]
    fn share_link_is_canonicalized_without_tracking() {
        let identity = parse_apple_music_playlist(
            "https://music.apple.com/hk/playlist/synthetic-list/pl.u-1234?l=en-GB&ls=1",
        )
        .expect("valid apple link");
        assert_eq!(identity.storefront, "hk");
        assert_eq!(identity.playlist_id, "pl.u-1234");
        assert!(!identity.public_url.contains('?'));
        assert!(parse_apple_music_playlist("https://example.com/hk/playlist/x/pl.u-1").is_err());
    }

    #[test]
    fn synthetic_catalog_tracks_are_provider_neutral() {
        let root = serde_json::from_value::<CatalogResponse>(serde_json::json!({
            "data": [{
                "id": "pl.u-1234",
                "type": "playlists",
                "attributes": {"name": "Synthetic", "url": "https://music.apple.com/hk/playlist/synthetic/pl.u-1234"},
                "relationships": {"tracks": {"data": [{
                    "id": "123456",
                    "type": "songs",
                    "attributes": {
                        "name": "Synthetic Song",
                        "artistName": "Synthetic Artist",
                        "albumName": "Synthetic Album",
                        "genreNames": ["Cantopop"],
                        "releaseDate": "2026-01-02",
                        "artwork": {"url": "https://is1-ssl.mzstatic.com/image/thumb/synthetic/{w}x{h}bb.{f}"},
                        "url": "https://music.apple.com/hk/song/synthetic/123456"
                    }
                }]}}
            }]
        }))
        .expect("fixture shape");
        let tracks = root
            .data
            .into_iter()
            .next()
            .and_then(|playlist| playlist.relationships)
            .and_then(|relationships| relationships.tracks)
            .expect("tracks")
            .data;
        let items = normalize_tracks(tracks, "hk").expect("normalize tracks");
        assert_eq!(items[0].source_provider, "apple-music");
        assert_eq!(items[0].genre, "Cantopop");
        assert!(items[0].cover_url.contains("300x300bb.jpg"));
    }

    #[test]
    fn next_page_cannot_escape_the_official_playlist_path() {
        assert!(
            validate_next_url(
                "/v1/catalog/hk/playlists/pl.u-1234/tracks?offset=100&limit=100",
                "hk",
                "pl.u-1234"
            )
            .is_ok()
        );
        assert!(validate_next_url("https://evil.example/tracks", "hk", "pl.u-1234").is_err());
        assert!(normalize_artwork_url("https://evil.example/{w}x{h}.{f}").is_err());
    }
}
